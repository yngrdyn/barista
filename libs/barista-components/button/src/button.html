<ng-content select="dt-loading-spinner"></ng-content>
<span class="dt-button-icon" *ngIf="_icons.length">
  <ng-content select="dt-icon"></ng-content>
</span>
<span class="dt-button-label">
  <ng-content></ng-content>
</span>
