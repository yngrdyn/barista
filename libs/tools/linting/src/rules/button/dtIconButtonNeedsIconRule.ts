/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { AttrAst, ElementAst } from '@angular/compiler';
import { BasicTemplateAstVisitor, NgWalker } from 'codelyzer';
import { IRuleMetadata, RuleFailure, Rules } from 'tslint';
import { SourceFile } from 'typescript';

import {
  addFailure,
  hasContent,
  hasOnlyDtIconChildren,
  isButtonElement,
  isIconButtonAttr,
} from '../../utils';

class DtButtonVisitor extends BasicTemplateAstVisitor {
  visitElement(element: ElementAst, context: any): void {
    this._validateElement(element);
    super.visitElement(element, context);
  }

  private _validateElement(element: ElementAst): any {
    if (!isButtonElement(element)) {
      return;
    }

    const attrs: AttrAst[] = element.attrs;
    const isIconButton = attrs.some((attr) => isIconButtonAttr(attr));

    if (isIconButton && !hasContent(element)) {
      addFailure(
        this,
        element,
        'A dt-icon-button must not be empty, but must contain a dt-icon element.',
      );
    }

    if (isIconButton) {
      if (hasOnlyDtIconChildren(element)) {
        return;
      }

      addFailure(
        this,
        element,
        'A dt-icon-button must contain dt-icon elements only. No other nested elements are allowed.',
      );
    }
  }
}

/**
 * The dtIconButtonNeedsIconRule ensures that an icon button only contains dt-icon elements.
 *
 * The following examples pass the button lint checks:
 * <button dt-icon-button variant="secondary"><dt-icon name="agent"></dt-icon></button>
 * <button dt-icon-button variant="primary">
 *  <dt-icon name="dont-watch" *ngIf="isExpanded"></dt-icon>
 *  <dt-icon name="overview" *ngIf="!isExpanded"></dt-icon>
 * </button>
 *
 * For the following example the linter throws an error:
 * <button dt-icon-button variant="secondary"><dt-icon name="agent"></dt-icon> icon button</button>, only dt-icon child elements allowed
 * <button dt-icon-button variant="secondary">icon button</button>, icon content required
 * <button dt-icon-button variant="secondary"></button>, icon content required
 */
export class Rule extends Rules.AbstractRule {
  static readonly metadata: IRuleMetadata = {
    description:
      'Ensures that an icon button contains only dt-icon components.',
    // tslint:disable-next-line:no-null-keyword
    options: null,
    optionsDescription: 'Not configurable.',
    rationale: 'An icon button must only contain dt-icon components.',
    ruleName: 'dt-icon-button-needs-icon',
    type: 'maintainability',
    typescriptOnly: true,
  };

  apply(sourceFile: SourceFile): RuleFailure[] {
    return this.applyWithWalker(
      new NgWalker(sourceFile, this.getOptions(), {
        templateVisitorCtrl: DtButtonVisitor,
      }),
    );
  }
}
