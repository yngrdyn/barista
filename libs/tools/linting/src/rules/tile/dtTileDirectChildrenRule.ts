/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ElementAst } from '@angular/compiler';
import { BasicTemplateAstVisitor, NgWalker } from 'codelyzer';
import { IRuleMetadata, RuleFailure, Rules } from 'tslint';
import { SourceFile } from 'typescript';

import {
  ChildNode,
  addFailure,
  findChild,
  isElementWithName,
} from '../../utils';

class DtTileVisitor extends BasicTemplateAstVisitor {
  visitElement(element: ElementAst, context: any): void {
    this._validateElement(element);
    super.visitElement(element, context);
  }

  private _validateElement(element: ElementAst): any {
    if (!isElementWithName(element, 'dt-tile')) {
      return;
    }

    const directChildren = [
      'dt-tile-title',
      'dt-tile-subtitle',
      'dt-tile-icon',
    ];

    let childNodes: ChildNode[] = [];
    directChildren.forEach(childName => {
      childNodes = childNodes.concat(findChild(element, childName, 0));
    });

    const filteredChildren: string[] = childNodes
      .filter(el => el.level > 1)
      .map(el => el.name);

    if (filteredChildren.length < 1) {
      return;
    }

    const childrenNames = Array.from(new Set(filteredChildren));

    addFailure(
      this,
      element,
      `The following elements must be direct children of a dt-tile: ${childrenNames.join(
        ', ',
      )}`,
    );
  }
}

/**
 * The dtTileDirectChildrenRule ensures that the defined elements are always direct children of a dt-tile.
 *
 * The following example passes the check:
 * <dt-tile>
 *   <dt-tile-icon><dt-icon name="agent"></dt-icon></dt-tile-icon>
 *   <dt-tile-title>L-W8-64-APMDay3</dt-tile-title>
 *   // ...
 * </dt-tile>
 *
 * For the following example the linter throws an error:
 * <dt-tile>
 *   <dt-tile-icon><dt-icon name="agent"></dt-icon></dt-tile-icon>
 *   <dt-tile-title>L-W8-64-APMDay3</dt-tile-title>
 *   <div>
 *     <dt-tile-subtitle>Linux (x84, 64-bit)</dt-tile-subtitle>
 *   </div>
 *   // ...
 * </dt-tile>
 */
export class Rule extends Rules.AbstractRule {
  static readonly metadata: IRuleMetadata = {
    description:
      "Ensures that a tile's child components are direct children of a dt-tile.",
    // tslint:disable-next-line:no-null-keyword
    options: null,
    optionsDescription: 'Not configurable.',
    rationale:
      "A tile's child components (title, subtitle, icon) must always be direct children of the dt-tile.",
    ruleName: 'dt-tile-direct-children',
    type: 'maintainability',
    typescriptOnly: true,
  };

  apply(sourceFile: SourceFile): RuleFailure[] {
    return this.applyWithWalker(
      new NgWalker(sourceFile, this.getOptions(), {
        templateVisitorCtrl: DtTileVisitor,
      }),
    );
  }
}
