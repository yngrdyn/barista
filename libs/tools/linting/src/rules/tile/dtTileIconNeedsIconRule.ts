/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ElementAst } from '@angular/compiler';
import { BasicTemplateAstVisitor, NgWalker } from 'codelyzer';
import { IRuleMetadata, RuleFailure, Rules } from 'tslint';
import { SourceFile } from 'typescript';

import {
  addFailure,
  hasContent,
  hasOnlyDtIconChildren,
  isElementWithName,
} from '../../utils';

class DtTileVisitor extends BasicTemplateAstVisitor {
  // Codelyzer in v6 has a dependency to angular 9 and therefore
  // we have to type the element as any, otherwise the installed
  // angular v10 and v9 conflict with their types
  visitElement(element: any, context: any): void {
    this._validateElement(element);
    super.visitElement(element, context);
  }

  private _validateElement(element: ElementAst): any {
    if (!isElementWithName(element, 'dt-tile-icon')) {
      return;
    }

    if (!hasContent(element)) {
      addFailure(
        this,
        element,
        'A dt-tile-icon must not be empty, but must contain a dt-icon element.',
      );
    }

    if (hasOnlyDtIconChildren(element)) {
      return;
    }

    addFailure(
      this,
      element,
      'A dt-tile-icon must contain dt-icon elements only. No other nested elements are allowed.',
    );
  }
}

/**
 * @deprecated Will be removed with v9.0.0 as tslint is deprecated and won't be supported anymore.
 * There will be no replacement for eslint. Instead take a look on our design system how the
 * component should be used.
 *
 * The dtTileIconNeedsIconRule ensures that a tile icon only contains dt-icon elements.
 *
 * The following example passes the lint checks:
 * <dt-tile-icon><dt-icon name="agent"></dt-icon></dt-tile-icon>
 *
 * For the following examples the linter throws an error:
 * <dt-tile-icon>some text</dt-tile-icon>
 * <dt-tile-icon> </dt-tile-icon>
 */
export class Rule extends Rules.AbstractRule {
  static readonly metadata: IRuleMetadata = {
    description: 'Ensures that a tile icon contains only dt-icon components.',
    // tslint:disable-next-line:no-null-keyword
    options: null,
    optionsDescription: 'Not configurable.',
    rationale: 'A tile icon must only contain dt-icon components.',
    ruleName: 'dt-tile-icon-needs-icon',
    type: 'maintainability',
    typescriptOnly: true,
  };

  apply(sourceFile: SourceFile): RuleFailure[] {
    return this.applyWithWalker(
      new NgWalker(sourceFile, this.getOptions(), {
        templateVisitorCtrl: DtTileVisitor,
      }),
    );
  }
}
