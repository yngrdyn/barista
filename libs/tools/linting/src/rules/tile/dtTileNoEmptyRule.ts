/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ElementAst } from '@angular/compiler';
import { BasicTemplateAstVisitor, NgWalker } from 'codelyzer';
import { IRuleMetadata, RuleFailure, Rules } from 'tslint';
import { SourceFile } from 'typescript';

import {
  addFailure,
  hasContentApartFrom,
  isElementWithName,
} from '../../utils';

class DtTileVisitor extends BasicTemplateAstVisitor {
  // Codelyzer in v6 has a dependency to angular 9 and therefore
  // we have to type the element as any, otherwise the installed
  // angular v10 and v9 conflict with their types
  visitElement(element: any, context: any): void {
    this._validateElement(element);
    super.visitElement(element, context);
  }

  private _validateElement(element: ElementAst): any {
    if (!isElementWithName(element, 'dt-tile')) {
      return;
    }

    const tileChildren = ['dt-tile-icon', 'dt-tile-title', 'dt-tile-subtitle'];

    if (hasContentApartFrom(element, tileChildren)) {
      return;
    }

    addFailure(
      this,
      element,
      'A dt-tile must always contain content apart from title, subtitle and icon.',
    );
  }
}

/**
 * @deprecated Will be removed with v9.0.0 as tslint is deprecated and won't be supported anymore.
 * There will be no replacement for eslint. Instead take a look on our design system how the
 * component should be used.
 *
 * The dtCardNoEmptyRule ensures that a dt-tile always contains content
 * apart from a title, subtitle and an icon.
 *
 * The following example passes the check:
 * <dt-tile>
 *   <dt-tile-icon><dt-icon name="agent"></dt-icon></dt-tile-icon>
 *   <dt-tile-title>L-W8-64-APMDay3</dt-tile-title>
 *   <dt-tile-subtitle>Linux (x84, 64-bit)</dt-tile-subtitle>
 *   Network traffic
 * </dt-tile>
 *
 * For the following example the linter throws an error:
 * <dt-tile>
 *   <dt-tile-icon><dt-icon name="agent"></dt-icon></dt-tile-icon>
 *   <dt-tile-title>L-W8-64-APMDay3</dt-tile-title>
 *   <dt-tile-subtitle>Linux (x84, 64-bit)</dt-tile-subtitle>
 * </dt-tile>
 */
export class Rule extends Rules.AbstractRule {
  static readonly metadata: IRuleMetadata = {
    description:
      'Ensures that a dt-tile always contains content apart from title, subtitle and icon.',
    // tslint:disable-next-line:no-null-keyword
    options: null,
    optionsDescription: 'Not configurable.',
    rationale:
      'A dt-tile must always contain content apart from title, subtitle and icon.',
    ruleName: 'dt-tile-no-empty',
    type: 'maintainability',
    typescriptOnly: true,
  };

  apply(sourceFile: SourceFile): RuleFailure[] {
    return this.applyWithWalker(
      new NgWalker(sourceFile, this.getOptions(), {
        templateVisitorCtrl: DtTileVisitor,
      }),
    );
  }
}
