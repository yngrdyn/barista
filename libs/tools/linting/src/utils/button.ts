/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { AttrAst, ElementAst } from '@angular/compiler';

export function isIconButtonAttr(attr: AttrAst): boolean {
  return attr.name === 'dt-icon-button';
}

export function isButtonAttr(attr: AttrAst): boolean {
  return attr.name === 'dt-button';
}

export function isButtonElement(element: ElementAst): boolean {
  const elementName = element.name;
  if (elementName !== 'button' && elementName !== 'a') {
    return false;
  }

  if (
    elementName === 'a' &&
    !element.attrs.some((attr) => isButtonAttr(attr) || isIconButtonAttr(attr))
  ) {
    return false;
  }

  return true;
}
