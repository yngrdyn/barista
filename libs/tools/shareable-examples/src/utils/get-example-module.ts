/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { sync } from 'glob';
import { basename, join, resolve } from 'path';
import { tsCreateSourceFile } from '@dynatrace/shared/node';
import {
  createPrinter,
  isImportDeclaration,
  isPropertyAssignment,
  isVariableStatement,
  Node,
  SourceFile,
  transform,
  TransformationContext,
  TransformerFactory,
  visitEachChild,
  visitNode,
  createIdentifier,
  PropertyAssignment,
  ImportDeclaration,
  isSpreadElement,
  createPropertyAssignment,
  isStringLiteral,
  isNamedImports,
  ImportSpecifier,
} from 'typescript';
import { ExampleAstFile } from './examples.interface';

/**
 * Creates a transformer to prepare the examples module, this includes the following:
 * - remove all imports from the module that are relative, except the currently processed example
 * - remove the DT_COMPONENT_EXAMPLES array declaration
 * - remove the `entryComponents` from the module decorator
 * - replace the `declarations` in the module decorator with the current example.
 */
export function createTransformer(
  exampleRoot: string,
  exampleClassName: string,
): TransformerFactory<SourceFile> {
  const transformer: TransformerFactory<SourceFile> = (
    context: TransformationContext,
  ) => {
    let importsFromExampleFile: string | undefined;

    function visitSpraedDeclarationAndReplaceIt(node: Node): Node | undefined {
      if (isSpreadElement(node)) {
        return importsFromExampleFile
          ? createIdentifier(importsFromExampleFile)
          : createIdentifier(exampleClassName);
      }
      return visitEachChild(node, visitSpraedDeclarationAndReplaceIt, context);
    }

    /** Visit a propertyAssignment and determine which assignments need to be rewritten. */
    function visitPropertyAssignment(
      node: PropertyAssignment,
    ): Node | undefined {
      const assignmentName = node.name.getText();
      // Replace the spread in the declarations with the one component
      // we want to keep in there.
      if (assignmentName === 'declarations') {
        return visitEachChild(
          node,
          visitSpraedDeclarationAndReplaceIt,
          context,
        );
      }
      // Replace the entryComponents assignment with an exports,
      // to be able to use the component outside of the declaring module.
      if (assignmentName === 'entryComponents') {
        const entryComponentsNode = visitEachChild(
          node,
          visitSpraedDeclarationAndReplaceIt,
          context,
        );
        return createPropertyAssignment(
          'exports',
          entryComponentsNode.initializer,
        );
      }
      return node;
    }

    /** Visit an importDeclaration and remove imports if necessary. */
    function visitImportDeclaration(node: ImportDeclaration): Node | undefined {
      const importFrom = isStringLiteral(node.moduleSpecifier)
        ? node.moduleSpecifier.text
        : '';
      // By definition, our examples are nested each within its own folder
      // we will need to keep the import for this one example, and all other
      // imports, that are not nested in a folder here.
      if (
        // Check if it is a relative import
        importFrom.startsWith('.') &&
        // Check if it is not the one import we actually want to keep
        !importFrom.includes(basename(exampleRoot)) &&
        // Check if it is not any of the other relative imports we want to ditch.
        importFrom.split('/').length > 2
      ) {
        return undefined;
      }
      // If the import is from the example component, remember the imports
      // as they will need to be in the declarations and definitions.
      if (
        importFrom.includes(basename(exampleRoot)) &&
        node.importClause &&
        node.importClause.namedBindings &&
        isNamedImports(node.importClause.namedBindings)
      ) {
        importsFromExampleFile = node.importClause.namedBindings.elements
          .map((element: ImportSpecifier) => element.name.text)
          .join(', ');
      }
      return node;
    }

    /** Main visit function. */
    function visit(node: Node): Node | undefined {
      // Remove all import declarations that are not needed anymore.
      if (isImportDeclaration(node)) {
        return visitNode(node, visitImportDeclaration);
      }

      // Remove the unnecessary array variable declaration
      if (isVariableStatement(node)) {
        // If the variable statement is the declarations array, remove it.
        if (/DT_(.*?)_EXAMPLES/.test(node.getText())) {
          return;
        }
      }

      // Replace the declarations and entryComponents assignments
      if (isPropertyAssignment(node)) {
        return visitNode(node, visitPropertyAssignment);
      }
      return visitEachChild(node, visit, context);
    }
    return node => visitNode(node, visit);
  };
  return transformer;
}

export async function getExampleModule(
  exampleRoot: string,
  exampleClassName: string,
): Promise<ExampleAstFile> {
  // Get the module file path
  // This assumes that the module is one level up the folder structure
  // and matches the pattern '.module.ts'
  const moduleFilePath = sync('*.module.ts', {
    cwd: resolve(exampleRoot, '../'),
  }).map(file => join('../', file))[0];
  const moduleAstFile = await tsCreateSourceFile(
    resolve(exampleRoot, moduleFilePath),
  );

  const result = transform(moduleAstFile, [
    createTransformer(exampleRoot, exampleClassName),
  ]);

  // To apply the changes the transformed file needs to be passed
  // to a printer to get the updated file.
  const printer = createPrinter();
  const content = printer.printFile(result.transformed[0]);
  return {
    path: join(exampleRoot, moduleFilePath),
    content,
    ast: result.transformed[0],
  };
}
