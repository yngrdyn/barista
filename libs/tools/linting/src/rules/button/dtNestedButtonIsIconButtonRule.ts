/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { AttrAst, ElementAst } from '@angular/compiler';
import { BasicTemplateAstVisitor, NgWalker } from 'codelyzer';
import { IRuleMetadata, RuleFailure, Rules } from 'tslint';
import { SourceFile } from 'typescript';

import { addFailure, isButtonElement, isIconButtonAttr } from '../../utils';

class DtButtonVisitor extends BasicTemplateAstVisitor {
  visitElement(element: ElementAst, context: any): void {
    this._validateElement(element);
    super.visitElement(element, context);
  }

  private _validateElement(element: ElementAst): any {
    if (!isButtonElement(element)) {
      return;
    }

    const attrs: AttrAst[] = element.attrs;
    const isNestedVariant = attrs.some(
      attr => attr.name === 'variant' && attr.value === 'nested',
    );
    const isIconButton = attrs.some(attr => isIconButtonAttr(attr));

    // dt-icon-button attribute required for nested buttons
    if (isNestedVariant && !isIconButton) {
      addFailure(
        this,
        element,
        'A nested button variant must always be a dt-icon-button.',
      );
    }
  }
}

/**
 * The dtNestedButtonIsIconButtonRule ensures that a nested button is always an icon-button.
 *
 * The following example passes the button lint checks:
 * <button dt-icon-button variant="nested"><dt-icon name="agent"></dt-icon></button>
 *
 * For the following example the linter throws errors:
 * <button dt-button variant="nested">...</button>, dt-icon-button attribute required
 */
export class Rule extends Rules.AbstractRule {
  static readonly metadata: IRuleMetadata = {
    description: 'Ensures that a nested button is always an icon button.',
    // tslint:disable-next-line:no-null-keyword
    options: null,
    optionsDescription: 'Not configurable.',
    rationale: 'A nested button must always be an icon button.',
    ruleName: 'dt-nested-button-is-icon-button',
    type: 'maintainability',
    typescriptOnly: true,
  };

  apply(sourceFile: SourceFile): RuleFailure[] {
    return this.applyWithWalker(
      new NgWalker(sourceFile, this.getOptions(), {
        templateVisitorCtrl: DtButtonVisitor,
      }),
    );
  }
}
