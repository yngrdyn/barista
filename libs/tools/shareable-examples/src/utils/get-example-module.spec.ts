/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { createTransformer } from './get-example-module';
import {
  createSourceFile,
  ScriptTarget,
  ScriptKind,
  SourceFile,
  transform,
  createPrinter,
} from 'typescript';

function transformFile(
  moduleSource: SourceFile,
  exampleRoot: string,
  exampleClassName: string,
): string {
  const result = transform(moduleSource, [
    createTransformer(exampleRoot, exampleClassName),
  ]);

  // To apply the changes the transformed file needs to be passed
  // to a printer to get the updated file.
  const printer = createPrinter();
  return printer.printFile(result.transformed[0]);
}

it('should remove all relative imports, which are not related to the current example', () => {
  const source = createSourceFile(
    'dummy.ts',
    `
import { NgModule } from '@angular/core';
import { DtOverlayModule } from '@dynatrace/barista-components/overlay';
import { DtKeyValueListModule } from '@dynatrace/barista-components/key-value-list';
import { DtButtonModule } from '@dynatrace/barista-components/button';
import { DtIconModule } from '@dynatrace/barista-components/icon';
import { DtTileModule } from '@dynatrace/barista-components/tile';

import { DtExampleOverlayComplexContent } from './overlay-complex-content-example/overlay-complex-content-example';
import { DtExampleOverlayDefault } from './overlay-default-example/overlay-default-example';
import { DtExampleOverlayImplicitContext } from './overlay-implicit-context-example/overlay-implicit-context-example';

    `,
    ScriptTarget.Latest,
    true, // setParentNodes
    ScriptKind.TS,
  );
  expect(
    transformFile(
      source,
      'barista-components/src/overlay-default-example/overlay-default-example',
      'DtExampleOverlayDefault',
    ),
  ).toMatchSnapshot();
});

it('should remove all relative imports, which are not related to the current example', () => {
  const source = createSourceFile(
    'dummy.ts',
    `
import { NgModule } from '@angular/core';
import { DtOverlayModule } from '@dynatrace/barista-components/overlay';
import { DtKeyValueListModule } from '@dynatrace/barista-components/key-value-list';
import { DtButtonModule } from '@dynatrace/barista-components/button';
import { DtIconModule } from '@dynatrace/barista-components/icon';
import { DtTileModule } from '@dynatrace/barista-components/tile';

import { SomeService } from './example.service.ts';
import { DtExampleOverlayComplexContent } from './overlay-complex-content-example/overlay-complex-content-example';
import { DtExampleOverlayDefault } from './overlay-default-example/overlay-default-example';
import { DtExampleOverlayImplicitContext } from './overlay-implicit-context-example/overlay-implicit-context-example';
    `,
    ScriptTarget.Latest,
    true, // setParentNodes
    ScriptKind.TS,
  );
  expect(
    transformFile(
      source,
      'barista-components/src/overlay-default-example/overlay-default-example',
      'DtExampleOverlayDefault',
    ),
  ).toMatchSnapshot();
});

it('should create declarations and exports in the decorator', () => {
  const source = createSourceFile(
    'dummy.ts',
    `
import { DtExampleChartAreaRange } from './chart-arearange-example/chart-arearange-example';
import { DtExampleChartBar } from './chart-bar-example/chart-bar-example';
import { DtExampleChartBehaviorSwitch } from './chart-behavior-switch-example/chart-behavior-switch-example';
import { DtExampleChartCategorized } from './chart-categorized-example/chart-categorized-example';
import { DtExampleChartDefault } from './chart-default-example/chart-default-example';
import { DtExampleChartDonut } from './chart-donut-example/chart-donut-example';
import { DtChartExampleDataService } from './chart-example-data.service';
import { DtExampleChartHeatfield } from './chart-heatfield-example/chart-heatfield-example';
import { DtExampleChartHeatfieldMultiple } from './chart-heatfield-multiple-example/chart-heatfield-multiple-example';
import { DtExampleChartLine } from './chart-line-example/chart-line-example';
import { DtExampleChartLineWithGaps } from './chart-line-with-gaps-example/chart-line-with-gaps-example';
import { DtExampleChartLoading } from './chart-loading-example/chart-loading-example';
import { DtExampleChartMinMax } from './chart-min-max-example/chart-min-max-example';
import { DtExampleChartOrderedColors } from './chart-ordered-colors-example/chart-ordered-colors-example';
import { DtExampleChartPie } from './chart-pie-example/chart-pie-example';
import { DtExampleChartSelectionAreaDefault } from './chart-selection-area-default-example/chart-selection-area-default-example';
import { DtExampleChartSinglePointData } from './chart-single-data-point-example/chart-single-data-point-example';
import { DtExampleChartStream } from './chart-stream-example/chart-stream-example';

@NgModule({
  imports: [
    CommonModule,
    DtChartModule,
    DtKeyValueListModule,
    DtButtonModule,
    DtButtonGroupModule,
  ],
  declarations: [
    DtExampleChartArea,
    DtExampleChartAreaRange,
    DtExampleChartBar,
    DtExampleChartBehaviorSwitch,
    DtExampleChartCategorized,
    DtExampleChartDefault,
    DtExampleChartDonut,
    DtExampleChartHeatfield,
    DtExampleChartHeatfieldMultiple,
    DtExampleChartLine,
    DtExampleChartLineWithGaps,
    DtExampleChartLoading,
    DtExampleChartMinMax,
    DtExampleChartOrderedColors,
    DtExampleChartPie,
    DtExampleChartSelectionAreaDefault,
    DtExampleChartSinglePointData,
    DtExampleChartStream,
  ],
  providers: [DtChartExampleDataService],
})
export class DtChartExamplesModule {}
    `,
    ScriptTarget.Latest,
    true, // setParentNodes
    ScriptKind.TS,
  );
  expect(
    transformFile(
      source,
      'barista-components/src/chart-min-max-example/chart-min-max-example',
      'DtExampleChartMinMax',
    ),
  ).toMatchSnapshot();
});
