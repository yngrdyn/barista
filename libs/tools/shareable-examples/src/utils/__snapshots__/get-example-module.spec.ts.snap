// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create declarations and exports in the decorator 1`] = `
"import { DtChartExampleDataService } from './chart-example-data.service';
import { DtExampleChartMinMax } from './chart-min-max-example/chart-min-max-example';
@NgModule({
    imports: [
        CommonModule,
        DtChartModule,
        DtKeyValueListModule,
        DtButtonModule,
        DtButtonGroupModule,
    ],
    declarations: [
        DtExampleChartMinMax,
    ],
    providers: [DtChartExampleDataService],
    exports: [DtExampleChartMinMax]
})
export class DtChartExamplesModule {
}
"
`;

exports[`should remove all relative imports, which are not related to the current example 1`] = `
"import { NgModule } from '@angular/core';
import { DtOverlayModule } from '@dynatrace/barista-components/overlay';
import { DtKeyValueListModule } from '@dynatrace/barista-components/key-value-list';
import { DtButtonModule } from '@dynatrace/barista-components/button';
import { DtIconModule } from '@dynatrace/barista-components/icon';
import { DtTileModule } from '@dynatrace/barista-components/tile';
import { DtExampleOverlayDefault } from './overlay-default-example/overlay-default-example';
"
`;

exports[`should remove all relative imports, which are not related to the current example 2`] = `
"import { NgModule } from '@angular/core';
import { DtOverlayModule } from '@dynatrace/barista-components/overlay';
import { DtKeyValueListModule } from '@dynatrace/barista-components/key-value-list';
import { DtButtonModule } from '@dynatrace/barista-components/button';
import { DtIconModule } from '@dynatrace/barista-components/icon';
import { DtTileModule } from '@dynatrace/barista-components/tile';
import { SomeService } from './example.service.ts';
import { DtExampleOverlayDefault } from './overlay-default-example/overlay-default-example';
"
`;
